src/mallocTracer.cpp

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright The async-profiler authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "asprof.h"
+#include "assert.h"
+#include "codeCache.h"
+#include "mallocTracer.h"
+#include "os.h"
+#include "profiler.h"
+#include <dlfcn.h>
+#include <string.h>
+
+#define ADDRESS_OF(sym) ({               \
+    void* addr = dlsym(RTLD_NEXT, #sym); \
+    addr != NULL ? (sym##_t)addr : sym;  \
+})
+
+typedef void* (*malloc_t)(size_t);
+static malloc_t _orig_malloc = NULL;
+
+typedef void* (*calloc_t)(size_t, size_t);
+static calloc_t _orig_calloc = NULL;
+
+typedef void* (*realloc_t)(void*, size_t);
+static realloc_t _orig_realloc = NULL;
+
+typedef void (*free_t)(void*);
+static free_t _orig_free = NULL;
+
+static void* malloc_hook(size_t size) {
+    void* ret = _orig_calloc(1, size);
+    if (likely(ret && size)) {
+        MallocTracer::recordMalloc(ret, size, OS::nanotime());
+    }
+    return ret;
+}
+
+#ifdef __linux__
+extern "C" WEAK DLLEXPORT void* malloc(size_t size) {
+    if (unlikely(!_orig_malloc)) {
+        return NULL;
+    }
+
+    if (likely(MallocTracer::initialized())) {
+        return malloc_hook(size);
+    }
+    return _orig_malloc(size);
+}
+#endif // __linux__
+
+static void* calloc_hook(size_t num, size_t size) {
+    void* ret = _orig_calloc(num, size);
+    if (likely(ret && num && size)) {
+        MallocTracer::recordMalloc(ret, num * size, OS::nanotime());
+    }
+    return ret;
+}
+
+#ifdef __linux__
+extern "C" WEAK DLLEXPORT void* calloc(size_t num, size_t size) {
+    if (unlikely(!_orig_calloc)) {
+        // In some libc versions, dlsym may call calloc on error case in the ADDRESS_OF macro.
+        return NULL;
```

Comment:
Does this really happen? Even if it does, I guess returning `NULL` will break something anyway.

Reply:
This could happen in the first `dlsym` from `.ctors`, which can call `calloc` in the tested libc implementation on an error case.

Returning NULL does not break anything as libc has a safe fallback. Libc will ignore failing `calloc` in this specific case and use another libc-internal variable instead.

Reply:
I thought you moved initialization of `_orig` pointer to a constructor exactly to solve the mentioned problem. If not, will a lazy resolution work the same way?

Reply:
There is a cycle we need to break: First `dlsym` call from `_orig_malloc = ADDRESS_OF(malloc)` can invoke calloc once. When `NULL` is returned, it can recover: https://github.com/bminor/glibc/blob/5daf0c98152bdc54719a3ef62302204705e16fa6/dlfcn/dlerror.c#L141

I fixed this in the new version by using `OS::safeAlloc` only while in `getOrigAddresses` method. Keeping returning `NULL` as it would be semantically correct and better than crashing when any `_orig_*` is NULL.